certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -867,3 +875,18 @@ def install_ssl_options_conf(options_ssl):
     # Check to make sure options-ssl.conf is installed
     if not os.path.isfile(options_ssl):
         shutil.copyfile(constants.MOD_SSL_CONF_SRC, options_ssl)
+        return
+    # there's already a file there. if it exactly matches a previous file hash,
+    # we can update it. otherwise, create a .new and print a warning.
+    digest = crypto_util.sha256sum(options_ssl)
+    if digest in PREVIOUS_SSL_OPTIONS_HASHES: # safe to update
+        shutil.copyfile(constants.MOD_SSL_CONF_SRC, options_ssl)
+    elif digest == CURRENT_SSL_OPTIONS_HASH: # already up to date
+        return
+    else: # not safe to update
+        new_filename = options_ssl + ".new"
+        shutil.copyfile(constants.MOD_SSL_CONF_SRC, new_filename)
+        # print warning
+        logger.warning("%s has been manually modified; updated ssl configuration options "
```

Comment:
One thing that's slightly awkward is if you run Certbot with the Nginx plugin once, modify `options-ssl-nginx.conf`, and then run it again, it tells you an updated file has been saved and you should update your modified version despite them being the same base file. If users are exposed to this behavior of Certbot telling you to update your file despite them being the same, I could see them adding `--nginx-no-tls-updates` just to get Certbot to shut up.

There are are number of ways to mitigate this, but basically I think we want to save `CURRENT_SSL_OPTIONS_HASH` (or the original file) somewhere and check if it's different than our current file before creating the `.new` file with a warning. `/var/lib/letsencrypt` would be nice, but then you have the slightly awkward fact that `options-ssl-nginx.conf` is stored in one configurable directory location and it's original hash/contents are stored in another. Perhaps we could just create a new (hidden) file/folder in `/etc/letsencrypt` to contain this?

This is definitely one downside to doing things at runtime that I didn't think of. If we feel this is too hard/complex to solve correctly, we could go back to your original plan and get things in a sane state for OS packages, `certbot-auto`, `docker`, (and `pip`?), but I still think that's more work and causes Certbot to have different behavior depending on how it's installed.